File: examples/go/09-new-features/pdfa_conversion/main.go
// PDF/A conversion — v0.3.40
//
// Demonstrates ConvertToPdfA: build a simple PDF, open it in the editor,
// convert to PDF/A-2b (level 2), then save the result.
// Run: go run -tags pdf_oxide_dev main.go

package main

import (
	"fmt"
	"log"
	"os"
	"path/filepath"

	pdfoxide "github.com/yfedoseev/pdf_oxide/go"
)

func must(err error) {
	if err != nil {
		log.Fatal(err)
	}
}

func main() {
	outDir := "output"
	must(os.MkdirAll(outDir, 0o755))

	// Build a simple PDF in memory.
	b, err := pdfoxide.NewDocumentBuilder()
	must(err)
	defer b.Close()
	must(b.Title("PDF/A Conversion Demo"))

	page, err := b.LetterPage()
	must(err)
	_, err = page.
		Font("Helvetica", 12).
		At(72, 720).
		Heading(1, "PDF/A-2b Conversion Demo").
		At(72, 690).
		Paragraph("This document will be converted to PDF/A-2b archival format.").
		Done()
	must(err)

	pdfBytes, err := b.Build()
	must(err)
	fmt.Printf("Original PDF size: %d bytes\n", len(pdfBytes))

	// Open with editor and convert to PDF/A-2b (level index 2).
	editor, err := pdfoxide.OpenEditorFromBytes(pdfBytes)
	must(err)
	defer editor.Close()

	// ConvertToPdfA: 0=A1b 1=A1a 2=A2b 3=A2a 4=A2u 5=A3b 6=A3a 7=A3u
	if err := editor.ConvertToPdfA(2); err != nil {
		// Non-fatal: log and continue — the save below still writes a valid PDF.
		fmt.Printf("PDF/A conversion note: %v\n", err)
	} else {
		fmt.Println("PDF/A-2b conversion succeeded")
	}

	result, err := editor.SaveToBytes()
	must(err)
	if len(result) == 0 {
		log.Fatal("saved PDF is empty")
	}
	fmt.Printf("Output PDF size: %d bytes\n", len(result))

	path := filepath.Join(outDir, "pdfa.pdf")
	must(os.WriteFile(path, result, 0o644))
	fmt.Printf("Written: %s\n", path)
}

File: examples/rust/09-new-features/pdfa_conversion/main.rs
// PDF/A conversion: validate → convert → validate — v0.3.41
//
// Demonstrates the full archival pipeline:
//   1. Build a PDF in memory
//   2. Validate PDF/A-2b conformance (expect errors before conversion)
//   3. Convert to PDF/A-2b in-place
//   4. Validate again (expect compliant or fewer errors)
//
// Run: cargo run --example showcase_pdfa_conversion

use pdf_oxide::{
    compliance::{convert_to_pdf_a, validate_pdf_a, PdfALevel},
    error::Result,
    writer::DocumentBuilder,
    PdfDocument,
};
use std::path::PathBuf;

fn main() -> Result<()> {
    let out_dir = PathBuf::from("target/examples_output/pdfa_conversion");
    std::fs::create_dir_all(&out_dir)?;

    let mut builder = DocumentBuilder::new();
    builder
        .letter_page()
        .font("Helvetica", 12.0)
        .at(72.0, 720.0)
        .heading(1, "PDF/A-2b Conversion Demo")
        .at(72.0, 690.0)
        .paragraph("This document will be converted to PDF/A-2b archival format.")
        .done();

    let pdf_bytes = builder.build()?;
    println!("Original PDF size: {} bytes", pdf_bytes.len());

    // Step 1: validate before conversion.
    let mut doc = PdfDocument::from_bytes(pdf_bytes.clone())?;
    let pre = validate_pdf_a(&mut doc, PdfALevel::A2b)?;
    println!("Before conversion — compliant: {}, errors: {}", pre.is_compliant, pre.errors.len());

    // Step 2: convert to PDF/A-2b.
    let result = convert_to_pdf_a(&mut doc, PdfALevel::A2b)?;
    println!("Conversion success: {}, actions: {}", result.success, result.actions.len());
    for action in &result.actions {
        println!("  - {:?}", action.action_type);
    }

    // Step 3: validate after conversion.
    let post = validate_pdf_a(&mut doc, PdfALevel::A2b)?;
    println!("After conversion  — compliant: {}, errors: {}", post.is_compliant, post.errors.len());

    let out_path = out_dir.join("pdfa_converted.pdf");
    std::fs::write(&out_path, &doc.source_bytes)?;
    println!("Written: {}", out_path.display());
    Ok(())
}
